### ISSUE TITLE
RFC: “What’s in a (NumPY) name?” — Changing PyTorch Operator Names

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#68174]
**Should we change the names of some popular PyTorch operators to make PyTorch more consistent with NumPy?**

Today in PyTorch there are operations — like repeat() —  that do one thing in PyTorch and another in NumPy. This is a surprising pain for NumPy users coming to PyTorch, and it means that regular users of both frameworks have to remember which ops are different.

Removing “stumbling blocks” like these is one of PyTorch’s major user experience goals. Over the past year and a half we’ve added over 100 operators from NumPy and SciPy, including the torch.fft, torch.linalg, and torch.special modules. These updates have made it easier than ever to transition between PyTorch and NumPy (and SciPy). Changing operator names is, however, a little disruptive since it requires users update their existing programs.

Here’s the operators this RFC proposes renaming:

* repeat → tile
    * PyTorch’s repeat is different from NumPy’s repeat but identical to NumPy’s tile
* split → split_size
    * PyTorch’s split is different from NumPy’s split because it asks what size the resulting chunks should be, instead of how many there should be (see this issue (https://github.com/pytorch/pytorch/issues/50012) for details)
* nonzero → argwhere
    * PyTorch’s nonzero is different from NumPy’s nonzero but identical to NumPy’s argwhere
* max → amax, max_with_indices
    * PyTorch’s max has unusual behavior where it sometimes returns just the maximum value of a tensor and other times returns the maximum and the indices of the maximum values. Splitting this into amax (which just returns the maximum) and max_with_indices (which always returns both) is a clear user experience.
* size → shape
    * PyTorch already supports the shape attribute, but we use the ambiguous “size” to refer to shape while other frameworks often use it to refer to numel. Encouraging more consistent use of shape and numel is clearer and more consistent with other frameworks.

To accommodate each rename the new operation would be added (in most cases above it already exists) and users would be pointed to the new name for two releases. Then we’d remove the old names for one release, and finally restore the old names with behavior consistent with NumPy. This means users will have half a year or longer to find and replace names in their programs.

What are your thoughts? Comment below if you think consistency with NumPy is important or you’d like to argue that these disruptions, while small, aren’t worthwhile.

**Comment by rgommers** [comment]
>     * PyTorch already supports the shape attribute, but we use the ambiguous “size” to refer to shape while other frameworks often use it to refer to numel. Encouraging more consistent use of shape and numel is clearer and more consistent with other frameworks.

Because I wasn't sure, let me point out that `Tensor.size` is a method, and `Tensor.shape` is an attribute - so `shape` is indeed compatible. `.numel` is also a method, which is annoying.

**Comment by rgommers** [comment]
Thanks for the proposal @mruberry. While I in principle like the idea of making everything 100% compatible, I think it'd be good to prioritize - some things are more important than others. Out of the proposed changes, here are the ones I think are most important:
1. `size` -> `shape`. The `.shape` attribute already works, so this doesn't seem particularly painful to start with - updating the docs and changing all internal code to use `.shape` is pain-free. And then at some point emitting a deprecation warning for `.size` would be very nice.
2. `nonzero` -> `argwhere`. Similar, add an alias and switch over; this is an often-used function, and the difference is annoying. Should not be too painful.

The `max` one I'm more ambivalent about, because while the current behavior is not great, neither are the new names. After making this change, it will still not match `np.max`.

`repeat` and `split` are functions that should ideally never be used. So I wouldn't spend too much bandwidth on this. Perhaps a better approach instead is to spend some cycles to let both PyTorch and NumPy adopt the `einops` primitives for more intuitive reshaping/repeating operations. That's be a very nice new feature, and 100% backwards compatible.

**Comment by rgommers** [comment]
Just to document, @pearu pointed out that it is in fact possible to make `.size` or `.numel` both an attribute and a method:
```python
>>> class A(int):
...     def __new__(cls, size, shape):
...         obj = int.__new__(cls, size)
...         obj._shape = shape
...         return obj
...     def __call__(self):
...         return self._shape
... 
>>> A(20, (4, 5))
20
>>> A(20, (4, 5))()
(4, 5)
>>> class MyTensor:
...     def __init__(self, shape):
...         self._size = A(shape[0] * shape[1], shape)
...     @property
...     def size(self):
...         return self._size
... 
>>> t = MyTensor((3, 4))
>>> t.size
12
>>> t.size()
(3, 4)
```
This is not necessarily pretty and hard to recommend, but it may come in handy to preserve backwards compatibility while also adding numpy and/or array API standard compatibility.

**Comment by pearu** [comment]
Here's how to apply the above idea to torch Tensor:
```python
>>> import torch
>>> t = torch.tensor([[1, 2], [3, 4]])
>>> # Current state:
...
>>> t.size
<built-in method size of Tensor object at 0x7fd8f037af90>
>>> t.size()
torch.Size([2, 2])
>>> # Apply size wrapper to torch.Tensor:
...
>>> class SizeWrapper(int):
...     def __new__(cls, tensor):
...         obj = int.__new__(cls, tensor.numel())
...         obj._shape = tensor.shape
...         return obj
...     def __call__(self): return self._shape
... 
>>> torch.Tensor.size = property(lambda self: SizeWrapper(self))
>>> # Now we can access size as an attribute or a method:
...
>>> t.size
4
>>> t.size()
torch.Size([2, 2])
```

**Comment by pmeier** [comment]
It should be

```python
class SizeWrapper(int):
    ...

    def __call__(self, dim=None):
        if dim is None:
            return self._shape

        return self._shape[dim]

**Comment by ezyang** [comment]
NGL, it's gonna be hard to rewire my brain to s/size/shape/

**Comment by mruberry** [comment]
> NGL, it's gonna be hard to rewire my brain to s/size/shape/

You'll get over it ;). "Shape" and "numel" are the truth.
